src/sources/util/tcp.rs

```diff
@@ -53,41 +54,27 @@ async fn make_listener(
         },
     }
 }
-pub trait IsErrorFatal {
-    fn is_error_fatal(&self) -> bool;
-}
-
-impl IsErrorFatal for LinesCodecError {
-    fn is_error_fatal(&self) -> bool {
-        false
-    }
-}
-
-impl IsErrorFatal for std::io::Error {
-    fn is_error_fatal(&self) -> bool {
-        true
-    }
-}
-
-pub trait TcpSource: Clone + Send + Sync + 'static
-where
-    <<Self as TcpSource>::Decoder as tokio_util::codec::Decoder>::Item: std::marker::Send,
-{
+pub trait TcpSource: Send + Sync + 'static {
     // Should be default: `std::io::Error`.
     // Right now this is unstable: https://github.com/rust-lang/rust/issues/29661
-    type Error: From<io::Error> + IsErrorFatal + std::fmt::Debug + std::fmt::Display + Send;
-    type Decoder: Decoder<Error = Self::Error> + Send + 'static + Send;
+    type Error: From<io::Error> + TcpError + std::fmt::Debug + std::fmt::Display + Send;
+    type Item: Into<Event> + Send;
+    type Decoder: Decoder<Item = (Self::Item, usize), Error = Self::Error> + Send + 'static;
 
-    fn decoder(&self) -> Self::Decoder;
+    fn build_decoder(&self) -> Self::Decoder;
 
-    fn build_event(&self, frame: <Self::Decoder as Decoder>::Item, host: Bytes) -> Option<Event>;
+    fn build_event(&self, item: Self::Item) -> Event {
+        item.into()
+    }
 
-    fn build_ack(&self, _frame: &<Self::Decoder as Decoder>::Item) -> Bytes {
+    fn handle_event(&self, event: &mut Event, host: Bytes, byte_size: usize);
+
+    fn build_ack(&self, _item: &Self::Item) -> Bytes {
         Bytes::new()
     }
 
     fn run(
-        self,
+        self: Arc<Self>,
```

Comment:
A comment on why we need `Arc<Self>` would be nice. Unless I'm missing something, it's not obvious from the implementation.

Reply:
This is a good question, and the reason for this has been hidden quite a few layers deep.

For the `TcpSource` to be `Clone`, `Decoder` needs to be clone. However, it stores `Box`es of a `Framer` and `Parser`. `Box` is not `Clone`, but can be made via `dyn_clone::DynClone`. However, that doesn't work for structs that store a `Box` with a generic type associated to the struct, and previously the `BytesDecoder` wrapper stored a generic type that was `Into<Bytes>`.

Using `Arc` instead of `Box` there also doesn't work there, because the inner `Decoder` needs to be mutated via `decode` calls and `Decoder` also needs to be `Send + Sync` because `TcpSource` can spawn multiple instances concurrently (one per request). I felt that adding `Arc<Mutex<_>>` was too much cost for an operation that is executed for every frame.

Therefore, I decided to use `Arc` for cloning the source and build a fresh decoder using `source.build_decoder` for each request.

---

Long story short: I went back to forcing `Decoder` to be `Clone` in 655ac4e64c79de39f6567bfe8a60f8c6fb2e0bb4 by identifying how to accomplish that. It required getting rid of `BytesDecoder` (which @lukesteensen also suggested due to implicit conversion into `Byte`), making the `Box`es `Clone` by using `dyn_clone::DynClone` and finally hacking in a `Clone` implementation for `tokio_util::codec::LengthDelimitedCodec` that builds it anew instead of cloning its internal state, because unfortunately it does not implement `Clone` even though all its internal fields do, see https://github.com/timberio/vector/blob/655ac4e64c79de39f6567bfe8a60f8c6fb2e0bb4/src/sources/util/decoding/framing/length_delimited.rs#L33-L40.